consensus/state.go

```diff
@@ -12,19 +12,47 @@ import (
 	"go.sia.tech/core/types"
 )
 
-const (
-	blocksPerYear = 144 * 365
-
-	hardforkDevAddr      = 10000
-	hardforkTax          = 21000
-	hardforkStorageProof = 100000
-	hardforkOak          = 135000
-	hardforkOakFix       = 139000
-	hardforkASIC         = 179000
-	hardforkFoundation   = 298000
-
-	foundationSubsidyFrequency = blocksPerYear / 12
-)
+type networkParams struct {
+	hardforkHeightDevAddr      uint64
+	hardforkHeightTax          uint64
+	hardforkHeightStorageProof uint64
+	hardforkHeightOak          uint64
+	hardforkHeightOakFix       uint64
+	hardforkHeightASIC         uint64
+	hardforkHeightFoundation   uint64
+
+	hardforkASICTotalTarget types.BlockID
+	hardforkASICTotalTime   time.Duration
+	minimumCoinbase         uint32 // in SC
+}
+
+var mainnetParams = &networkParams{
+	hardforkHeightDevAddr:      10000,
+	hardforkHeightTax:          21000,
+	hardforkHeightStorageProof: 100000,
+	hardforkHeightOak:          135000,
+	hardforkHeightOakFix:       139000,
+	hardforkHeightASIC:         179000,
+	hardforkHeightFoundation:   298000,
+
+	hardforkASICTotalTarget: types.BlockID{8: 32},
+	hardforkASICTotalTime:   120000 * time.Second,
+	minimumCoinbase:         30000,
+}
+
+var testnetParams = &networkParams{
+	hardforkHeightDevAddr:      10000,
+	hardforkHeightTax:          2,
+	hardforkHeightStorageProof: 100000,
```

Comment:
`hardforkHeightDevAddr` and `hardforkHeightStorageProof` are currently hard-coded in `modules/consensus`. Ideally `siad` would define them properly in `types.go` so that we could specify testnet heights for them. Of course, that would require resetting the testnet, so maybe not worth it?

Reply:
I think fixing those two consts and resetting the testnet is a good idea

Reply:
Looks like the dev address hardfork should be disabled on the testnet
```go
if build.Release != "testnet" && sfo.UnlockHash == gsa[len(gsa)-1].UnlockHash && blockHeight(tx) > 10e3 {
    sfo.UnlockHash = devAddr
}
```

Since this resets the initial siafund allocation to a hardcoded address, I'm not sure what the best way to have it activate on the testnet is. We'd likely want to remove the hardcoded `devAddr` too.

Reply:
I would argue that siad's behavior is faulty here; it's using `gsa[len(gsa)-1]` as shorthand for the (old) dev address, but really it should be hardcoding it (which is what `core` does). If we hardcode it to the mainnet addr, it won't affect testnet at all.

Reply:
The testnet already overrides the proof hardfork height too. I added to the hardcoded if statement instead of defining a new variable 😬 
```go
if (build.Release == "standard" && height < 100e3) || (build.Release == "testnet" && height < 5) || (build.Release == "testing" && height < 10) {
    return validStorageProofs100e3(tx, t)
}
```

Reply:
We shouldn't need to reset the testnet even if we fix the logic in `siad`.

Reply:
siad changes are in SiaFoundation/siad#156